Various mypy fixes

holmanb opened the pull request:
## Additional Context
See individual commits.

This work enables using more of mypy's static analysis features.

This PR does various cleanups including:

- remove unnecessary type ignore comments
- add missing type annotations
- remove invalid type annotations
- minor refactors to eliminate the need for type annotations
- eliminate unused code
- auto-format network jsonschema
- tox: fix minor issues
- tox: add stub libraries


Note to reviewers:

Most of the changes here are no-ops. That said, please pay special attention to the commit titled `refactor(typing): Remove unused code paths`. Invalid type annotations can lead mypy to warn incorrectly that a code path is unused (this is how I found many of the type errors in this commit series). I do believe that I was thorough in verifying this commit, but this one carries the most risk of the series.

TheRealFalcon reviewed (commented):
I don't have time to do a full review today, but I realized I have a stale "Pending" comment that is still relevant.

holmanb commented on cloudinit/config/cc_power_state_change.py:
```diff
@@ -124,7 +124,7 @@ def check_condition(cond):
 
 
-def handle(name: str, cfg: Config, cloud: Cloud, args: list) -> None:
+def handle(name: str, cfg: Config, cloud: Cloud, _) -> None:
```
> This is part of every single handler...

This is the only one that has an argument name which later has a different type. If you'd rather we didn't fix it this way, then I'll change the shadowed variable name.

TheRealFalcon replied:
Ah, I see now. I thought you were doing something else. Yeah, I'm ok changing this one here. I think I'd still prefer the name change though.

TheRealFalcon commented on cloudinit/net/netplan.py:
```diff
@@ -482,9 +479,18 @@ def _render_content(self, network_state: NetworkState) -> str:
                 bridge_ports = ifcfg.get("bridge_ports")
-                # mypy wrong error. `copy(None)` is supported:
-                ports = sorted(copy.copy(bridge_ports))  # type: ignore
+                if bridge_ports is None:
```
Is there a reason we're downgrading to a warning here? I don't think we should be changing unrelated functionality in this kind of PR.

holmanb replied:
This was incorrectly silencing a mypy error which is valid.

```python
>>> sorted(copy.copy(None))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object is not iterable
```

holmanb replied:
The current code's comment is incorrect, and I suspect we could handle this situation more gracefully.

The current code prevents iteration from continuing. The intent here was to allow iteration to continue while also logging a warning, which I think is a more graceful way to handle invalid network state than throwing an exception which prevents the code from iterating through other interfaces.

TheRealFalcon replied:
+1. At first I thought we would want to throw an exception if we have issues here as to not render invalid config, but warning makes sense.

TheRealFalcon commented on cloudinit/util.py:
```diff
@@ -3231,8 +3229,8 @@ def deprecate(
     """
-    if not hasattr(deprecate, "_log"):
-        deprecate._log = set()  # type: ignore
+    if not deprecate.__dict__.get("log"):
```
[modifying __dict__ directly is not recommended](https://docs.python.org/3/library/stdtypes.html#modules). Why not `getattr` and `setattr` here, further down in this file, and in conftest.py?

TheRealFalcon commented on cloudinit/config/schemas/schema-network-config-v2.json:
```diff
@@ -10,43 +10,43 @@
-            "type": "boolean",
-            "description": "Unsupported  for dhcp6-overrides when used with the networkd renderer."
-          }
+      "type": "object",
```
Don't need to change in this PR, but separate functionality should really go in separate PRs. Multiple commits make sense in a PR when the functionality in a later commit requires the change from the earlier commit, but lets try to keep PRs are small and focused as possible.

holmanb replied:
+1, my intention with this was to make this part of the mypy + jsonschema type generation stuff, for which this makes more sense, but without I guess it feels like a bolt-on PR without that work.

TheRealFalcon reviewed (changes requested):
Comments inline

holmanb commented on setup_utils.py:
```diff
@@ -35,15 +35,6 @@ def pkg_config_read(library: str, var: str) -> str:
     return path
 
 
-def in_virtualenv() -> bool:
```
Let me know what needs to be added to the commit message:
```
  commit 5b0d2fa1bf714d7a952a27a59f1ce070159ff352
  Author: Brett Holman <brett.holman@canonical.com>
  Date:   Fri May 31 12:56:50 2024 -0600
  
      ci(mypy): Set default follow_imports value (#5350)
      
      Remove code for detecting a virtualenv. This has been broken since ~3.7.
  
   pyproject.toml |  1 -
   setup.py       | 17 +++++++----------
   setup_utils.py |  9 ---------
   3 files changed, 7 insertions(+), 20 deletions(-)
```

holmanb replied:
Mypy correctly points out that this isn't used anywhere on 3.12. Checking for the presence of a virtualenv hasn't worked since Python 3.7, and since nobody has bothered to make it work on >3.7, the assumption is that nobody cares about it working on 3.7 either. This appears to be a relic of the past.

Other factors: 

- We planned to drop support for 3.7 a year ago. We held out for a few stragglers that still wanted support, but I think it is time for 3.7 to go.
- Lots of disruption is coming in the packaging space in the upcoming cycle for PEP517. This is just packaging code. If someone still really wants this, they can maintain it themselves.

Aside: Earlier this year when I was trying to work on PEP517 support in my spare time, this code was actually a stumbling block due to the added complexity. Culling unnecessary junk prior to a big refactor or change in functionality helps a lot in the long run to understand what we actually need.

TheRealFalcon replied:
> Mypy correctly points out that this isn't used anywhere on 3.12

Didn't realize that. I don't disagree with anything you said. I knew this needed to go; I just didn't see the relation to this PR.

> 3.7

Let's start with 3.6 :wink: 

According to our original proposed calendar, we should still be supporting 3.7 until the end of the month.

holmanb commented on cloudinit/url_helper.py:
```diff
@@ -278,8 +278,6 @@ def __init__(self, response: requests.Response):
 
     @property
     def contents(self) -> bytes:
-        if self._response.content is None:
```
I'll need to dig in further about this one

holmanb replied:
@TheRealFalcon Thanks for finding that. It turns out that this was due to a bug in typeshed. I just [submitted a PR to fix it](https://github.com/python/typeshed/pull/12180). I'll change this PR to silence this warning until a fix is released.

holmanb commented:
@TheRealFalcon I think I've addressed all comments. If tests pass, I think this is ready for review.

TheRealFalcon commented on cloudinit/util.py:
```diff
@@ -3231,8 +3229,8 @@ def deprecate(
     """
-    if not hasattr(deprecate, "_log"):
-        deprecate._log = set()  # type: ignore
+    if not hasattr(deprecate, "log"):
```
Looking at this again, I don't think we want this (or the test updates based on it).

See the first comment on https://stackoverflow.com/a/16955098/37205

TheRealFalcon commented on .git-blame-ignore-revs:
```diff
@@ -2,3 +2,7 @@
 bae9b11da9ed7dd0b16fe5adeaf4774b7cc628cf
+
+# Apply json formatting
+04c82eafd8fe34c5f02ffecfdb5dba888f3baa3f
```
These changed again. Think it might make sense to update these after the fact?

TheRealFalcon commented on tox.ini:
```diff
@@ -78,7 +82,9 @@ deps =
     types-setuptools=={[format_deps]types-setuptools}
     typing-extensions=={[format_deps]typing-extensions}
-commands = {envpython} -m mypy cloudinit/ tests/ tools/
+# TODO: add --check-untyped-defs to allow checking untyped function bodies
```
This is done as of https://github.com/canonical/cloud-init/pull/5409 , correct?

holmanb replied:
Yes!

TheRealFalcon reviewed (changes requested):
I left a few more inline. The logging one is the only major one.

holmanb commented:
Looks like [the recent addition](355560312461caab01e914bb2424c6cd8268c153) to mypy features expanded the scope slightly, so I had to update some test code to make it pass. I think I just need to fix the merge conflict before this is ready to merge @TheRealFalcon.

TheRealFalcon commented on tests/integration_tests/instances.py:
```diff
@@ -60,7 +59,7 @@ class IntegrationInstance:
         self,
         cloud: "IntegrationCloud",
-        instance: BaseInstance,
+        instance,
```
This was fairly useful when it comes to editor autocomplete. Why does mypy not like it?

TheRealFalcon commented on tests/integration_tests/instances.py:
```diff
@@ -283,20 +282,6 @@ def _apt_update(self):
                     f"Failed to update packages: {response.stderr}"
                 )
 
-    def ip(self) -> str:
```
This is used in `__exit__` (though the implementation is buggy). Why is it being removed?

holmanb replied:
The whole purpose of that was to to make it easier for someone to ssh into an instance on a public cloud after a test, however since only lxd has `execute_via_ssh`, its purpose is defeated. I've been meaning to clean this up for a while, but hadn't gotten to it yet. Mypy is complaining about the types in this code which is why I finally made the change.

TheRealFalcon replied:
So can we fix it instead of dropping it? It's valid if we don't want to fix it here, but we can use the `# typing: ignore` directives in that case.

holmanb replied:
Sure, that's fair. Looking at this again I think the fix might be easier than I previously thought.

TheRealFalcon reviewed (commented):
2 questions inline

holmanb commented:
@TheRealFalcon Just force pushed to fix the merge conflict, and while I was at it I squashed the fixup commits and updated commit messages with PR number. Assuming tests pass this one is ready for re-review.

TheRealFalcon commented on tests/unittests/conftest.py:
```diff
@@ -152,7 +152,7 @@ def clear_deprecation_log():
     # a deprecation warning in a previous test can cause the next test to
     # fail.
-    util.deprecate._log = set()
+    setattr(util.deprecate, "log", set())
```
Are you sure this doesn't change any functionality? If so, +1, but I just want to make sure.

TheRealFalcon reviewed (approved):
One question inline, but otherwise LGTM!
